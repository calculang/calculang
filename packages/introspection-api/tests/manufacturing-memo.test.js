import introspection from '@calculang/introspection-api';

// TODO I don't expect this to work for modular models yet because of dependence on -nomemo run first !

//jest.setTimeout(20000);
jest.setTimeout(2000000); // for debug!

test.each([
  'manufacturing/base',
    // Disabling modular tests, existing memo implementation is unsupported here
  /*
  'manufacturing/price-change-reconciliation',
  'manufacturing/revenue-fixed-inputs',
  'manufacturing/impactsAB',
  'manufacturing/revenue-with-demand-curve',*/
  'manufacturing/simple-neg-A',
  'manufacturing/simple-neg-B',
  'manufacturing/simple-neg-C',
])('%s', async (d) => {
  const output = await introspection(
    `./packages/calculang-testcase-models/${d}.cul.js`,
    { memo: true }
  );

  expect({
    cul_scope_ids_to_resource: output.cul_scope_ids_to_resource,
    import_sources_to_resource: output.import_sources_to_resource,
    cul_input_map: output.cul_input_map,
    cul_functions: output.cul_functions,
    cul_links: output.cul_links,
  }).toMatchSnapshot();
});

// memo test is that base, with memoization on, should match memo-test with memoization off

test('introspection of base with memo on matches memo-test with memo off', async () => {
  const base_memo_on = await introspection(
    `./packages/calculang-testcase-models/manufacturing/base.cul.js`,
    { memo: true }
  );
  const test_memo_off = await introspection(
    `./packages/calculang-testcase-models/manufacturing/memo-test.cul.js`,
    {}
  );

  const sort_fn_text = (a, b) => { // from https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/sort
    const nameA = a.name; //a.name.toUpperCase(); // ignore upper and lowercase
    const nameB = b.name; //b.name.toUpperCase(); // ignore upper and lowercase
    if (nameA < nameB) {
      return -1;
    }
    if (nameA > nameB) {
      return 1;
    }
  
    // names must be equal
    return 0;
  };

  let base_memo_on_cul_functions_no_locs = [...base_memo_on.cul_functions.values()].sort(sort_fn_text).map(({loc, ...d}) => (d));
  let test_memo_off_cul_functions_no_locs = [...test_memo_off.cul_functions.values()].sort(sort_fn_text).map(({loc, ...d}) => (d)); // poss ordering prob in comparison?

  expect(base_memo_on.cul_links).toEqual(test_memo_off.cul_links);
  expect(base_memo_on.cul_input_map).toEqual(test_memo_off.cul_input_map);
  expect(base_memo_on_cul_functions_no_locs).toEqual(test_memo_off_cul_functions_no_locs);
});

// should disable? disabling
/*
test('introspection of revenue-fixed-inputs with memo on', async () => {
  const base_memo_on = await introspection(
    `./packages/calculang-testcase-models/manufacturing/revenue-fixed-inputs.cul.js`,
    { memo: true }
  );

  expect(base_memo_on).toMatchSnapshot();
});*/

// this is affected by a lot of things, including overrides-of-overrides problem
/*test('introspection of price-change-reconciliation with memo on', async () => {
  const base_memo_on = await introspection(
    `./packages/calculang-testcase-models/manufacturing/price-change-reconciliation.cul.js`,
    { memo: true }
  );

  expect(base_memo_on.cul_links).toMatchSnapshot();
});
*/
// dupe to facilitate debugging
/*test('introspection of impactsAB with memo on', async () => {
  const base_memo_on = await introspection(
    `./packages/calculang-testcase-models/manufacturing/impactsAB.cul.js`,
    { memo: true }
  );

  //expect(base_memo_on.cul_links).toMatchSnapshot();
});
*/