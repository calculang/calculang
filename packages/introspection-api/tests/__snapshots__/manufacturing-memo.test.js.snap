// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`manufacturing/base 1`] = `
Object {
  "cul_functions": Map {
    "0_revenue$" => Object {
      "cul_scope_id": 0,
      "cul_source_scope_id": 1,
      "imported": "revenue_",
      "inputs": Array [
        "units_in",
        "price_in",
      ],
      "name": "revenue$",
      "reason": "explicit import",
    },
    "0_costs$" => Object {
      "cul_scope_id": 0,
      "cul_source_scope_id": 1,
      "imported": "costs_",
      "inputs": Array [
        "units_in",
      ],
      "name": "costs$",
      "reason": "explicit import",
    },
    "0_profit$" => Object {
      "cul_scope_id": 0,
      "cul_source_scope_id": 1,
      "imported": "profit_",
      "inputs": Array [
        "units_in",
        "price_in",
      ],
      "name": "profit$",
      "reason": "explicit import",
    },
    "0_units$" => Object {
      "cul_scope_id": 0,
      "cul_source_scope_id": 1,
      "imported": "units_",
      "inputs": Array [
        "units_in",
      ],
      "name": "units$",
      "reason": "explicit import",
    },
    "0_price$" => Object {
      "cul_scope_id": 0,
      "cul_source_scope_id": 1,
      "imported": "price_",
      "inputs": Array [
        "price_in",
      ],
      "name": "price$",
      "reason": "explicit import",
    },
    "0_revenue" => Object {
      "cul_scope_id": 0,
      "cul_source_scope_id": 0,
      "inputs": Array [
        "units_in",
        "price_in",
      ],
      "name": "revenue",
      "reason": "definition",
    },
    "0_costs" => Object {
      "cul_scope_id": 0,
      "cul_source_scope_id": 0,
      "inputs": Array [
        "units_in",
      ],
      "name": "costs",
      "reason": "definition",
    },
    "0_profit" => Object {
      "cul_scope_id": 0,
      "cul_source_scope_id": 0,
      "inputs": Array [
        "units_in",
        "price_in",
      ],
      "name": "profit",
      "reason": "definition",
    },
    "0_units" => Object {
      "cul_scope_id": 0,
      "cul_source_scope_id": 0,
      "inputs": Array [
        "units_in",
      ],
      "name": "units",
      "reason": "definition",
    },
    "0_price" => Object {
      "cul_scope_id": 0,
      "cul_source_scope_id": 0,
      "inputs": Array [
        "price_in",
      ],
      "name": "price",
      "reason": "definition",
    },
    "1_revenue" => Object {
      "cul_scope_id": 1,
      "cul_source_scope_id": 0,
      "inputs": Array [
        "units_in",
        "price_in",
      ],
      "name": "revenue",
      "reason": "implicit import",
    },
    "1_costs" => Object {
      "cul_scope_id": 1,
      "cul_source_scope_id": 0,
      "inputs": Array [
        "units_in",
      ],
      "name": "costs",
      "reason": "implicit import",
    },
    "1_profit" => Object {
      "cul_scope_id": 1,
      "cul_source_scope_id": 0,
      "inputs": Array [
        "units_in",
        "price_in",
      ],
      "name": "profit",
      "reason": "implicit import",
    },
    "1_units" => Object {
      "cul_scope_id": 1,
      "cul_source_scope_id": 0,
      "inputs": Array [
        "units_in",
      ],
      "name": "units",
      "reason": "implicit import",
    },
    "1_price" => Object {
      "cul_scope_id": 1,
      "cul_source_scope_id": 0,
      "inputs": Array [
        "price_in",
      ],
      "name": "price",
      "reason": "implicit import",
    },
    "1_revenue_" => Object {
      "cul_scope_id": 1,
      "cul_source_scope_id": 1,
      "inputs": Array [
        "units_in",
        "price_in",
      ],
      "name": "revenue_",
      "reason": "definition (renamed)",
    },
    "1_costs_" => Object {
      "cul_scope_id": 1,
      "cul_source_scope_id": 1,
      "inputs": Array [
        "units_in",
      ],
      "name": "costs_",
      "reason": "definition (renamed)",
    },
    "1_profit_" => Object {
      "cul_scope_id": 1,
      "cul_source_scope_id": 1,
      "inputs": Array [
        "units_in",
        "price_in",
      ],
      "name": "profit_",
      "reason": "definition (renamed)",
    },
    "1_units_" => Object {
      "cul_scope_id": 1,
      "cul_source_scope_id": 1,
      "inputs": Array [
        "units_in",
      ],
      "name": "units_",
      "reason": "definition (renamed)",
    },
    "1_units_in" => Object {
      "cul_scope_id": 1,
      "cul_source_scope_id": 1,
      "inputs": Array [
        "units_in",
      ],
      "name": "units_in",
      "reason": "input definition",
    },
    "1_price_" => Object {
      "cul_scope_id": 1,
      "cul_source_scope_id": 1,
      "inputs": Array [
        "price_in",
      ],
      "name": "price_",
      "reason": "definition (renamed)",
    },
    "1_price_in" => Object {
      "cul_scope_id": 1,
      "cul_source_scope_id": 1,
      "inputs": Array [
        "price_in",
      ],
      "name": "price_in",
      "reason": "input definition",
    },
  },
  "cul_input_map": Map {
    "0_revenue$" => Set {
      "units_in",
      "price_in",
    },
    "0_costs$" => Set {
      "units_in",
    },
    "0_profit$" => Set {
      "units_in",
      "price_in",
    },
    "0_units$" => Set {
      "units_in",
    },
    "0_price$" => Set {
      "price_in",
    },
    "0_revenue" => Set {
      "units_in",
      "price_in",
    },
    "0_costs" => Set {
      "units_in",
    },
    "0_profit" => Set {
      "units_in",
      "price_in",
    },
    "0_units" => Set {
      "units_in",
    },
    "0_price" => Set {
      "price_in",
    },
    "1_revenue" => Set {
      "units_in",
      "price_in",
    },
    "1_costs" => Set {
      "units_in",
    },
    "1_profit" => Set {
      "units_in",
      "price_in",
    },
    "1_units" => Set {
      "units_in",
    },
    "1_price" => Set {
      "price_in",
    },
    "1_revenue_" => Set {
      "units_in",
      "price_in",
    },
    "1_costs_" => Set {
      "units_in",
    },
    "1_profit_" => Set {
      "units_in",
      "price_in",
    },
    "1_units_" => Set {
      "units_in",
    },
    "1_units_in" => Set {
      "units_in",
    },
    "1_price_" => Set {
      "price_in",
    },
    "1_price_in" => Set {
      "price_in",
    },
  },
  "cul_links": Set {
    Object {
      "from": "1_revenue_",
      "reason": "explicit import",
      "to": "0_revenue$",
    },
    Object {
      "from": "1_costs_",
      "reason": "explicit import",
      "to": "0_costs$",
    },
    Object {
      "from": "1_profit_",
      "reason": "explicit import",
      "to": "0_profit$",
    },
    Object {
      "from": "1_units_",
      "reason": "explicit import",
      "to": "0_units$",
    },
    Object {
      "from": "1_price_",
      "reason": "explicit import",
      "to": "0_price$",
    },
    Object {
      "from": "0_revenue$",
      "negs": Array [],
      "reason": "call",
      "to": "0_revenue",
    },
    Object {
      "from": "0_costs$",
      "negs": Array [],
      "reason": "call",
      "to": "0_costs",
    },
    Object {
      "from": "0_profit$",
      "negs": Array [],
      "reason": "call",
      "to": "0_profit",
    },
    Object {
      "from": "0_units$",
      "negs": Array [],
      "reason": "call",
      "to": "0_units",
    },
    Object {
      "from": "0_price$",
      "negs": Array [],
      "reason": "call",
      "to": "0_price",
    },
    Object {
      "from": "0_revenue",
      "reason": "implicit import",
      "to": "1_revenue",
    },
    Object {
      "from": "0_costs",
      "reason": "implicit import",
      "to": "1_costs",
    },
    Object {
      "from": "0_profit",
      "reason": "implicit import",
      "to": "1_profit",
    },
    Object {
      "from": "0_units",
      "reason": "implicit import",
      "to": "1_units",
    },
    Object {
      "from": "0_price",
      "reason": "implicit import",
      "to": "1_price",
    },
    Object {
      "from": "1_units",
      "negs": Array [],
      "reason": "call",
      "to": "1_revenue_",
    },
    Object {
      "from": "1_price",
      "negs": Array [],
      "reason": "call",
      "to": "1_revenue_",
    },
    Object {
      "from": "1_units",
      "negs": Array [],
      "reason": "call",
      "to": "1_costs_",
    },
    Object {
      "from": "1_revenue",
      "negs": Array [],
      "reason": "call",
      "to": "1_profit_",
    },
    Object {
      "from": "1_costs",
      "negs": Array [],
      "reason": "call",
      "to": "1_profit_",
    },
    Object {
      "from": "1_units_in",
      "reason": "input",
      "to": "1_units_",
    },
    Object {
      "from": "1_price_in",
      "reason": "input",
      "to": "1_price_",
    },
  },
  "cul_scope_ids_to_resource": Map {
    0 => "./base.cul.js",
    1 => "./base.cul.js?+memoed&cul_scope_id=1&cul_parent_scope_id=0",
  },
  "import_sources_to_resource": Map {
    "0_./base.cul.js?+memoed" => "./base.cul.js?+memoed&cul_scope_id=1&cul_parent_scope_id=0",
  },
}
`;

exports[`manufacturing/simple-neg-A 1`] = `
Object {
  "cul_functions": Map {
    "0_sales_by_product$" => Object {
      "cul_scope_id": 0,
      "cul_source_scope_id": 1,
      "imported": "sales_by_product_",
      "inputs": Array [
        "product_in",
      ],
      "name": "sales_by_product$",
      "reason": "explicit import",
    },
    "0_product$" => Object {
      "cul_scope_id": 0,
      "cul_source_scope_id": 1,
      "imported": "product_",
      "inputs": Array [
        "product_in",
      ],
      "name": "product$",
      "reason": "explicit import",
    },
    "0_sales_total$" => Object {
      "cul_scope_id": 0,
      "cul_source_scope_id": 1,
      "imported": "sales_total_",
      "inputs": Array [
        "product_in",
      ],
      "name": "sales_total$",
      "reason": "explicit import",
    },
    "0_sales_by_product" => Object {
      "cul_scope_id": 0,
      "cul_source_scope_id": 0,
      "inputs": Array [
        "product_in",
      ],
      "name": "sales_by_product",
      "reason": "definition",
    },
    "0_product" => Object {
      "cul_scope_id": 0,
      "cul_source_scope_id": 0,
      "inputs": Array [
        "product_in",
      ],
      "name": "product",
      "reason": "definition",
    },
    "0_sales_total" => Object {
      "cul_scope_id": 0,
      "cul_source_scope_id": 0,
      "inputs": Array [
        "product_in",
      ],
      "name": "sales_total",
      "reason": "definition",
    },
    "1_sales_by_product" => Object {
      "cul_scope_id": 1,
      "cul_source_scope_id": 0,
      "inputs": Array [
        "product_in",
      ],
      "name": "sales_by_product",
      "reason": "implicit import",
    },
    "1_product" => Object {
      "cul_scope_id": 1,
      "cul_source_scope_id": 0,
      "inputs": Array [
        "product_in",
      ],
      "name": "product",
      "reason": "implicit import",
    },
    "1_sales_total" => Object {
      "cul_scope_id": 1,
      "cul_source_scope_id": 0,
      "inputs": Array [
        "product_in",
      ],
      "name": "sales_total",
      "reason": "implicit import",
    },
    "1_sales_by_product_" => Object {
      "cul_scope_id": 1,
      "cul_source_scope_id": 1,
      "inputs": Array [
        "product_in",
      ],
      "name": "sales_by_product_",
      "reason": "definition (renamed)",
    },
    "1_product_" => Object {
      "cul_scope_id": 1,
      "cul_source_scope_id": 1,
      "inputs": Array [
        "product_in",
      ],
      "name": "product_",
      "reason": "definition (renamed)",
    },
    "1_product_in" => Object {
      "cul_scope_id": 1,
      "cul_source_scope_id": 1,
      "inputs": Array [
        "product_in",
      ],
      "name": "product_in",
      "reason": "input definition",
    },
    "1_sales_total_" => Object {
      "cul_scope_id": 1,
      "cul_source_scope_id": 1,
      "inputs": Array [
        "product_in",
      ],
      "name": "sales_total_",
      "reason": "definition (renamed)",
    },
  },
  "cul_input_map": Map {
    "0_sales_by_product$" => Set {
      "product_in",
    },
    "0_product$" => Set {
      "product_in",
    },
    "0_sales_total$" => Set {
      "product_in",
    },
    "0_sales_by_product" => Set {
      "product_in",
    },
    "0_product" => Set {
      "product_in",
    },
    "0_sales_total" => Set {
      "product_in",
    },
    "1_sales_by_product" => Set {
      "product_in",
    },
    "1_product" => Set {
      "product_in",
    },
    "1_sales_total" => Set {
      "product_in",
    },
    "1_sales_by_product_" => Set {
      "product_in",
    },
    "1_product_" => Set {
      "product_in",
    },
    "1_product_in" => Set {
      "product_in",
    },
    "1_sales_total_" => Set {
      "product_in",
    },
  },
  "cul_links": Set {
    Object {
      "from": "1_sales_by_product_",
      "reason": "explicit import",
      "to": "0_sales_by_product$",
    },
    Object {
      "from": "1_product_",
      "reason": "explicit import",
      "to": "0_product$",
    },
    Object {
      "from": "1_sales_total_",
      "reason": "explicit import",
      "to": "0_sales_total$",
    },
    Object {
      "from": "0_sales_by_product$",
      "negs": Array [],
      "reason": "call",
      "to": "0_sales_by_product",
    },
    Object {
      "from": "0_product$",
      "negs": Array [],
      "reason": "call",
      "to": "0_product",
    },
    Object {
      "from": "0_sales_total$",
      "negs": Array [],
      "reason": "call",
      "to": "0_sales_total",
    },
    Object {
      "from": "0_sales_by_product",
      "reason": "implicit import",
      "to": "1_sales_by_product",
    },
    Object {
      "from": "0_product",
      "reason": "implicit import",
      "to": "1_product",
    },
    Object {
      "from": "0_sales_total",
      "reason": "implicit import",
      "to": "1_sales_total",
    },
    Object {
      "from": "1_product",
      "negs": Array [],
      "reason": "call",
      "to": "1_sales_by_product_",
    },
    Object {
      "from": "1_product_in",
      "reason": "input",
      "to": "1_product_",
    },
    Object {
      "from": "1_sales_by_product",
      "negs": Array [
        "product_in",
      ],
      "reason": "call",
      "to": "1_sales_total_",
    },
    Object {
      "from": "1_sales_by_product",
      "negs": Array [],
      "reason": "call",
      "to": "1_sales_total_",
    },
  },
  "cul_scope_ids_to_resource": Map {
    0 => "./simple-neg-A.cul.js",
    1 => "./simple-neg-A.cul.js?+memoed&cul_scope_id=1&cul_parent_scope_id=0",
  },
  "import_sources_to_resource": Map {
    "0_./simple-neg-A.cul.js?+memoed" => "./simple-neg-A.cul.js?+memoed&cul_scope_id=1&cul_parent_scope_id=0",
  },
}
`;

exports[`manufacturing/simple-neg-B 1`] = `
Object {
  "cul_functions": Map {
    "0_sales_by_product$" => Object {
      "cul_scope_id": 0,
      "cul_source_scope_id": 1,
      "imported": "sales_by_product_",
      "inputs": Array [
        "product_in",
      ],
      "name": "sales_by_product$",
      "reason": "explicit import",
    },
    "0_product$" => Object {
      "cul_scope_id": 0,
      "cul_source_scope_id": 1,
      "imported": "product_",
      "inputs": Array [
        "product_in",
      ],
      "name": "product$",
      "reason": "explicit import",
    },
    "0_sales_total$" => Object {
      "cul_scope_id": 0,
      "cul_source_scope_id": 1,
      "imported": "sales_total_",
      "inputs": Array [],
      "name": "sales_total$",
      "reason": "explicit import",
    },
    "0_sales_by_product" => Object {
      "cul_scope_id": 0,
      "cul_source_scope_id": 0,
      "inputs": Array [
        "product_in",
      ],
      "name": "sales_by_product",
      "reason": "definition",
    },
    "0_product" => Object {
      "cul_scope_id": 0,
      "cul_source_scope_id": 0,
      "inputs": Array [
        "product_in",
      ],
      "name": "product",
      "reason": "definition",
    },
    "0_sales_total" => Object {
      "cul_scope_id": 0,
      "cul_source_scope_id": 0,
      "inputs": Array [],
      "name": "sales_total",
      "reason": "definition",
    },
    "1_sales_by_product" => Object {
      "cul_scope_id": 1,
      "cul_source_scope_id": 0,
      "inputs": Array [
        "product_in",
      ],
      "name": "sales_by_product",
      "reason": "implicit import",
    },
    "1_product" => Object {
      "cul_scope_id": 1,
      "cul_source_scope_id": 0,
      "inputs": Array [
        "product_in",
      ],
      "name": "product",
      "reason": "implicit import",
    },
    "1_sales_total" => Object {
      "cul_scope_id": 1,
      "cul_source_scope_id": 0,
      "inputs": Array [],
      "name": "sales_total",
      "reason": "implicit import",
    },
    "1_sales_by_product_" => Object {
      "cul_scope_id": 1,
      "cul_source_scope_id": 1,
      "inputs": Array [
        "product_in",
      ],
      "name": "sales_by_product_",
      "reason": "definition (renamed)",
    },
    "1_product_" => Object {
      "cul_scope_id": 1,
      "cul_source_scope_id": 1,
      "inputs": Array [
        "product_in",
      ],
      "name": "product_",
      "reason": "definition (renamed)",
    },
    "1_product_in" => Object {
      "cul_scope_id": 1,
      "cul_source_scope_id": 1,
      "inputs": Array [
        "product_in",
      ],
      "name": "product_in",
      "reason": "input definition",
    },
    "1_sales_total_" => Object {
      "cul_scope_id": 1,
      "cul_source_scope_id": 1,
      "inputs": Array [],
      "name": "sales_total_",
      "reason": "definition (renamed)",
    },
  },
  "cul_input_map": Map {
    "0_sales_by_product$" => Set {
      "product_in",
    },
    "0_product$" => Set {
      "product_in",
    },
    "0_sales_total$" => Set {},
    "0_sales_by_product" => Set {
      "product_in",
    },
    "0_product" => Set {
      "product_in",
    },
    "0_sales_total" => Set {},
    "1_sales_by_product" => Set {
      "product_in",
    },
    "1_product" => Set {
      "product_in",
    },
    "1_sales_total" => Set {},
    "1_sales_by_product_" => Set {
      "product_in",
    },
    "1_product_" => Set {
      "product_in",
    },
    "1_product_in" => Set {
      "product_in",
    },
    "1_sales_total_" => Set {},
  },
  "cul_links": Set {
    Object {
      "from": "1_sales_by_product_",
      "reason": "explicit import",
      "to": "0_sales_by_product$",
    },
    Object {
      "from": "1_product_",
      "reason": "explicit import",
      "to": "0_product$",
    },
    Object {
      "from": "1_sales_total_",
      "reason": "explicit import",
      "to": "0_sales_total$",
    },
    Object {
      "from": "0_sales_by_product$",
      "negs": Array [],
      "reason": "call",
      "to": "0_sales_by_product",
    },
    Object {
      "from": "0_product$",
      "negs": Array [],
      "reason": "call",
      "to": "0_product",
    },
    Object {
      "from": "0_sales_total$",
      "negs": Array [],
      "reason": "call",
      "to": "0_sales_total",
    },
    Object {
      "from": "0_sales_by_product",
      "reason": "implicit import",
      "to": "1_sales_by_product",
    },
    Object {
      "from": "0_product",
      "reason": "implicit import",
      "to": "1_product",
    },
    Object {
      "from": "0_sales_total",
      "reason": "implicit import",
      "to": "1_sales_total",
    },
    Object {
      "from": "1_product",
      "negs": Array [],
      "reason": "call",
      "to": "1_sales_by_product_",
    },
    Object {
      "from": "1_product_in",
      "reason": "input",
      "to": "1_product_",
    },
    Object {
      "from": "1_sales_by_product",
      "negs": Array [
        "product_in",
      ],
      "reason": "call",
      "to": "1_sales_total_",
    },
    Object {
      "from": "1_sales_by_product",
      "negs": Array [
        "product_in",
      ],
      "reason": "call",
      "to": "1_sales_total_",
    },
  },
  "cul_scope_ids_to_resource": Map {
    0 => "./simple-neg-B.cul.js",
    1 => "./simple-neg-B.cul.js?+memoed&cul_scope_id=1&cul_parent_scope_id=0",
  },
  "import_sources_to_resource": Map {
    "0_./simple-neg-B.cul.js?+memoed" => "./simple-neg-B.cul.js?+memoed&cul_scope_id=1&cul_parent_scope_id=0",
  },
}
`;

exports[`manufacturing/simple-neg-C 1`] = `
Object {
  "cul_functions": Map {
    "0_sales_by_product$" => Object {
      "cul_scope_id": 0,
      "cul_source_scope_id": 1,
      "imported": "sales_by_product_",
      "inputs": Array [
        "product_in",
      ],
      "name": "sales_by_product$",
      "reason": "explicit import",
    },
    "0_product$" => Object {
      "cul_scope_id": 0,
      "cul_source_scope_id": 1,
      "imported": "product_",
      "inputs": Array [
        "product_in",
      ],
      "name": "product$",
      "reason": "explicit import",
    },
    "0_sales_total$" => Object {
      "cul_scope_id": 0,
      "cul_source_scope_id": 1,
      "imported": "sales_total_",
      "inputs": Array [],
      "name": "sales_total$",
      "reason": "explicit import",
    },
    "0_sales_by_product" => Object {
      "cul_scope_id": 0,
      "cul_source_scope_id": 0,
      "inputs": Array [
        "product_in",
      ],
      "name": "sales_by_product",
      "reason": "definition",
    },
    "0_product" => Object {
      "cul_scope_id": 0,
      "cul_source_scope_id": 0,
      "inputs": Array [
        "product_in",
      ],
      "name": "product",
      "reason": "definition",
    },
    "0_sales_total" => Object {
      "cul_scope_id": 0,
      "cul_source_scope_id": 0,
      "inputs": Array [],
      "name": "sales_total",
      "reason": "definition",
    },
    "1_sales_by_product" => Object {
      "cul_scope_id": 1,
      "cul_source_scope_id": 0,
      "inputs": Array [
        "product_in",
      ],
      "name": "sales_by_product",
      "reason": "implicit import",
    },
    "1_product" => Object {
      "cul_scope_id": 1,
      "cul_source_scope_id": 0,
      "inputs": Array [
        "product_in",
      ],
      "name": "product",
      "reason": "implicit import",
    },
    "1_sales_total" => Object {
      "cul_scope_id": 1,
      "cul_source_scope_id": 0,
      "inputs": Array [],
      "name": "sales_total",
      "reason": "implicit import",
    },
    "1_sales_by_product_" => Object {
      "cul_scope_id": 1,
      "cul_source_scope_id": 1,
      "inputs": Array [
        "product_in",
      ],
      "name": "sales_by_product_",
      "reason": "definition (renamed)",
    },
    "1_product_" => Object {
      "cul_scope_id": 1,
      "cul_source_scope_id": 1,
      "inputs": Array [
        "product_in",
      ],
      "name": "product_",
      "reason": "definition (renamed)",
    },
    "1_product_in" => Object {
      "cul_scope_id": 1,
      "cul_source_scope_id": 1,
      "inputs": Array [
        "product_in",
      ],
      "name": "product_in",
      "reason": "input definition",
    },
    "1_sales_total_" => Object {
      "cul_scope_id": 1,
      "cul_source_scope_id": 1,
      "inputs": Array [],
      "name": "sales_total_",
      "reason": "definition (renamed)",
    },
  },
  "cul_input_map": Map {
    "0_sales_by_product$" => Set {
      "product_in",
    },
    "0_product$" => Set {
      "product_in",
    },
    "0_sales_total$" => Set {},
    "0_sales_by_product" => Set {
      "product_in",
    },
    "0_product" => Set {
      "product_in",
    },
    "0_sales_total" => Set {},
    "1_sales_by_product" => Set {
      "product_in",
    },
    "1_product" => Set {
      "product_in",
    },
    "1_sales_total" => Set {},
    "1_sales_by_product_" => Set {
      "product_in",
    },
    "1_product_" => Set {
      "product_in",
    },
    "1_product_in" => Set {
      "product_in",
    },
    "1_sales_total_" => Set {},
  },
  "cul_links": Set {
    Object {
      "from": "1_sales_by_product_",
      "reason": "explicit import",
      "to": "0_sales_by_product$",
    },
    Object {
      "from": "1_product_",
      "reason": "explicit import",
      "to": "0_product$",
    },
    Object {
      "from": "1_sales_total_",
      "reason": "explicit import",
      "to": "0_sales_total$",
    },
    Object {
      "from": "0_sales_by_product$",
      "negs": Array [],
      "reason": "call",
      "to": "0_sales_by_product",
    },
    Object {
      "from": "0_product$",
      "negs": Array [],
      "reason": "call",
      "to": "0_product",
    },
    Object {
      "from": "0_sales_total$",
      "negs": Array [],
      "reason": "call",
      "to": "0_sales_total",
    },
    Object {
      "from": "0_sales_by_product",
      "reason": "implicit import",
      "to": "1_sales_by_product",
    },
    Object {
      "from": "0_product",
      "reason": "implicit import",
      "to": "1_product",
    },
    Object {
      "from": "0_sales_total",
      "reason": "implicit import",
      "to": "1_sales_total",
    },
    Object {
      "from": "1_product",
      "negs": Array [],
      "reason": "call",
      "to": "1_sales_by_product_",
    },
    Object {
      "from": "1_product_in",
      "reason": "input",
      "to": "1_product_",
    },
    Object {
      "from": "1_sales_by_product",
      "negs": Array [
        "product_in",
      ],
      "reason": "call",
      "to": "1_sales_total_",
    },
  },
  "cul_scope_ids_to_resource": Map {
    0 => "./simple-neg-C.cul.js",
    1 => "./simple-neg-C.cul.js?+memoed&cul_scope_id=1&cul_parent_scope_id=0",
  },
  "import_sources_to_resource": Map {
    "0_./simple-neg-C.cul.js?+memoed" => "./simple-neg-C.cul.js?+memoed&cul_scope_id=1&cul_parent_scope_id=0",
  },
}
`;
